### PR TITLE
220 profile banner attachment

### LINKED ISSUE #220: [220] Profile banner attachment
**Current state**
There is no way to add a profile banner and/or crop your banner.

**Expected state**
The user should be able to crop their profile banner and also change it by uploading a file to our server. There are constraints on the uploaded files which have been made by previous issues.

This issue should be made in the same way as the following issues:
#208 
#211 

**TO DO**
- [x] Upload image (Drag 'n drop & clickable)
- [x] Crop image
- [x] Show image on profile load

**Necessities**
#208 
#211 

**Branch**
220-profile-banner-attachment

### PR DESCRIPTION
**Pull request by jerohero**
closes #220 

 **What is it supposed to do**
Allows users to update their banners by cropping a given image to the required format.

 **How can we test your branch**
Navigate to your profile while logged in. Click on the banner (a floating action button should pop up as well). Try out some images. Maybe try out some images with odd sizes (such as very wide or tall images) as well if you have those. Try to crop it and save it. It should be updated immediately, and this banner should be shown again when you refresh te page.

Merging develop with this branch caused a lot of merge conflicts (hence the rather long delay), so it may be possible I destroyed some other functionalities that I did not notice. The functionalities that caused trouble in the merge conflicts are updating profile avatars and following users. So maybe try to see if it still works (works for me though).

 **What broweser have been checked**
- [x] Chrome
- [X] Firefox
- [x] Opera
- [X] Edge
- [ ] Safari

### CODE REVIEW

File: backend/app/controller/profileController.ts [hunk 1]
```diff
@@ -136,10 +129,12 @@ class ProfileController {
...
             description: profile.description
         }
-        const attachment = await this.dao.getProfileAttachment(profile.id)
+        const attachments = await this.dao.getProfileAttachments(profile.id)
 
-        if (attachment)
-            payload['avatar'] = 'http://localhost:8000/' + attachment.path
+        if ( attachments.avatar )
+            payload['avatar'] = 'http://localhost:8000/' + attachments.avatar.path
+        if ( attachments.banner )
```
Comment (AjobK) [inline backend/app/controller/profileController.ts]: newline above

File: backend/app/controller/profileController.ts [hunk 2]
```diff
@@ -241,6 +236,47 @@ class ProfileController {
...
+        const profile = await this.dao.getProfileByUsername( username )
+        const attachments = await this.dao.getProfileAttachments(profile.id)
+        const location = await this.fileService.storeImage(file, type)
+
+        let attachment = attachments[type]
+
+        // Create type dependent variables
+        let dimensions
+        let typeDefaultPath = 'default/'
+        if (type === AVATAR) {
```
Comment (AjobK) [inline backend/app/controller/profileController.ts]: newline above

File: backend/app/controller/profileController.ts [hunk 3]
```diff
@@ -241,6 +236,47 @@ class ProfileController {
...
     }
 
+    private updateAttachment = async (username, file, type): Promise<any> => {
+        const profile = await this.dao.getProfileByUsername( username )
+        const attachments = await this.dao.getProfileAttachments(profile.id)
+        const location = await this.fileService.storeImage(file, type)
+
+        let attachment = attachments[type]
+
+        // Create type dependent variables
```
Comment (AjobK) [inline backend/app/controller/profileController.ts]: remove

File: backend/app/controller/profileController.ts [hunk 4]
```diff
@@ -241,6 +236,47 @@ class ProfileController {
...
+
+        if (attachment.path !== typeDefaultPath) {
+            this.fileService.deleteImage(attachment.path)
+            attachment.path = location
+
+            return await this.attachmentDAO.saveAttachment(attachment)
+        }
+
+        let typeField
+        if (type === AVATAR)
```
Comment (AjobK) [inline backend/app/controller/profileController.ts]: newline above

File: backend/app/dao/attachmentDAO.ts [hunk 5]
```diff
@@ -11,10 +11,18 @@ class attachmentDAO {
...
     public async getAttachment(id: number): Promise<Attachment> {
         const repository = await DatabaseConnector.getRepository('Attachment')
-        const attachment = await repository.findOne({ where: { id: id } })
-
+        const attachment = await repository.findOne({where: { id: id }})
         return attachment
     }
+    public async getDefaultAvatarAttachment(): Promise<Attachment> {
+        const repository = await DatabaseConnector.getRepository('Attachment')
+        return await repository.findOne({where: {id: 1}})
```
Comment (AjobK) [inline backend/app/dao/attachmentDAO.ts]: newline above

File: backend/app/dao/attachmentDAO.ts [hunk 6]
```diff
@@ -11,10 +11,18 @@ class attachmentDAO {
...
+    }
+    public async getDefaultBannerAttachment(): Promise<Attachment> {
+        const repository = await DatabaseConnector.getRepository('Attachment')
+        return await repository.findOne({where: {id: 2}})
```
Comment (AjobK) [inline backend/app/dao/attachmentDAO.ts]: newline above

File: backend/app/dao/profileDAO.ts [hunk 7]
```diff
@@ -10,11 +10,14 @@ class ProfileDAO {
...
 
-    public async getProfileAttachment(profileId: number): Promise<Attachment> {
+    public async getProfileAttachments(profileId: number): Promise<{ avatar: Attachment, banner: Attachment }> {
         const repositoryProfile = await DatabaseConnector.getRepository('Profile')
-        const profile = await repositoryProfile.findOne({ where: { id: profileId }, relations: ['avatar_attachment'] })
-
-        return !profile.avatar_attachment ? null : profile.avatar_attachment
+        const avatar = await repositoryProfile.findOne({ where: { id: profileId }, relations: ['avatar_attachment'] })
+        const banner = await repositoryProfile.findOne({ where: { id: profileId }, relations: ['banner_attachment'] })
+        return {
```
Comment (AjobK) [inline backend/app/dao/profileDAO.ts]: newline above

File: backend/app/util/fileService.ts [hunk 8]
```diff
@@ -43,8 +45,7 @@ class FileService {
 
         newPath = newPath + '/' + month + '/' + day + '/' + name
         await pipeline(file.stream, fs.createWriteStream(newPath))
-
-        return 'profile/' + year + '/' + month + '/' + day + '/' + name
+        return 'profile/' + type + '/' + year + '/' + month + '/' + day + '/' + name
```
Comment (AjobK) [inline backend/app/util/fileService.ts]: newline above

File: frontend/src/components/cropper/index.js [hunk 9]
```diff
@@ -1,23 +1,37 @@
...
+                aspect: 16/9
+            }
+        } else if (props.inputType === this.AVATAR) {
+            crop = {
                 unit: '%',
                 width: 100,
                 aspect: 1
-            },
+            }
+        }
```
Comment (AjobK) [inline frontend/src/components/cropper/index.js]: Why not

```
let crop = {
    unit: '%',
    width: 100,
    aspect: 1
}

if (props.inputType === this.BANNER) {
    crop.aspect = 16/9
}
```

File: frontend/src/components/cropper/index.js [hunk 10]
```diff
@@ -98,42 +110,52 @@ class AvatarUpload extends Component {
...
-            if (err.response.status === 401) {
-                this.props.history.push('/login/')
-            }
-        })
+        let address = this.props.store.defaultData.backendUrl
+        if (this.props.inputType === this.BANNER) {
+            address += '/profile/banner'
+        } else if (this.props.inputType === this.AVATAR) {
+            address += '/profile/picture'
+        }
```
Comment (AjobK) [inline frontend/src/components/cropper/index.js]: Option 1 (Better)
```suggestion
        const { inputType } = this.props
        let address = this.props.store.defaultData.backendUrl + '/profile/'
        
        if (inputType === this.BANNER) {
            address += 'banner'
        } else if (inputType === this.AVATAR) {
            address += 'picture'
        }
```

Would probably have been better to just use this.BANNER of this.AVATAR which would result in the following:

Option 2 (I think this is even better)
```suggestion
        const { inputType } = this.props
        let address = this.props.store.defaultData.backendUrl + '/profile/' + inputType
```

File: frontend/src/components/cropper/index.js [hunk 11]
```diff
@@ -98,42 +110,52 @@ class AvatarUpload extends Component {
...
+        Axios.put(address,
+            fd,
+            {withCredentials: true, 'content-type': 'multipart/form-data'})
+            .then((res) => {
+                this.props.changeImage(res.data.url) // put in Axios response
+                this.props.closeCropper()
+            })
+            .catch(err => {
+                console.log(err)
+                if (err.response.status === 401) {
```
Comment (AjobK) [inline frontend/src/components/cropper/index.js:134]: newline above

File: frontend/src/components/cropper/index.js [hunk 12]
```diff
@@ -98,42 +110,52 @@ class AvatarUpload extends Component {
...
+        let address = this.props.store.defaultData.backendUrl
+        if (this.props.inputType === this.BANNER) {
+            address += '/profile/banner'
+        } else if (this.props.inputType === this.AVATAR) {
+            address += '/profile/picture'
+        }
+
+        Axios.put(address,
+            fd,
+            {withCredentials: true, 'content-type': 'multipart/form-data'})
```
Comment (AjobK) [inline frontend/src/components/cropper/index.js]:
```suggestion
            {
                'withCredentials': true,
                'content-type': 'multipart/form-data'
            }
        )
```

File: frontend/src/components/userBanner/index.js [hunk 13]
```diff
@@ -11,47 +11,73 @@ class UserBanner extends Component {
...
+      draggingOverAvatar: false
+    })
+  }
+
+  onBannerDragEnter = () => {
+    this.setState({
+      draggingOverBanner: true
+    })
+  }
+  onBannerDragLeave = () => {
```
Comment (AjobK) [inline frontend/src/components/userBanner/index.js:65]: newline above

**Review by AjobK — CHANGES_REQUESTED** [review]
Changes based on viewed code

**Review by AjobK — CHANGES_REQUESTED** [review]
Automatically create folder on saving banner or avatar. Avatar doesnt work either...

**Comment by AjobK** [comment]
image

Close popup when clicking outside of it. Scrolls when holding inside the white bars on mobile?

**Comment by jerohero** [comment]
Can't get updating avatars and banners to work anymore. Even in develop it suddenly seems to be broken and I have no idea why. The frontend is properly sending the file formdata, yet the backend doesn't recognize it.
image
Req.file is null and the middleware causes a crash:
image
This isn't my code, so maybe @ryankroon00 can take a look at it?

**Comment by AjobK** [comment]
@jerohero We will discuss this at tomorrows meeting since we're working until atleast 15:00 anyway

File: backend/app/controller/profileController.ts [hunk 14]
```diff
@@ -241,6 +237,48 @@ class ProfileController {
         return res.status(200).json({ message: `Succesfully ${followedProfile ? '' : 'un'}followed profile`, following: followedProfile })
     }
 
+    private updateAttachment = async (username, file, type): Promise<any> => {
+        const profile = await this.dao.getProfileByUsername( username )
+        const attachments = await this.dao.getProfileAttachments(profile.id)
+        const location = await this.fileService.storeImage(file, type)
+
+        console.log(attachments)
```
Comment (AjobK) [inline backend/app/controller/profileController.ts]:
```suggestion
```

**Review by AjobK — CHANGES_REQUESTED** [review]

**Comment by AjobK** [comment]
@jerohero 

Check merge conflicts

File: .gitignore [hunk 15]
```diff
@@ -38,4 +38,7 @@ package-lock.json
 backend/node_modules
 backend/ormconfig.js
 backend/ormconfig.json
-backend/app/public/profile/*
+backend/app/public/profile/banner/*
+backend/app/public/profile/avatar/*
+!backend/app/public/profile/banner/.gitkeep
+!backend/app/public/profile/avatar/.gitkeep
```
Comment (AjobK) [inline .gitignore]: newline

File: backend/app/public/profile/avatar/.gitkeep [hunk 16]
```diff
@@ -0,0 +1 @@
+this file exists so the empty avatar folder gets pushed
```
Comment (AjobK) [inline backend/app/public/profile/avatar/.gitkeep]: newline

File: backend/app/public/profile/banner/.gitkeep [hunk 17]
```diff
@@ -0,0 +1 @@
+this file exists so the empty banner folder gets pushed
```
Comment (AjobK) [inline backend/app/public/profile/banner/.gitkeep]: newline

File: backend/app/public/profile/avatar/.gitkeep (hunk 16 above)
Comment (AjobK) [reply backend/app/public/profile/avatar/.gitkeep]:
```suggestion
This file exists so the empty avatar folder gets pushed.
```

File: backend/app/public/profile/banner/.gitkeep (hunk 17 above)
Comment (AjobK) [reply backend/app/public/profile/banner/.gitkeep]:
```suggestion
This file exists so the empty banner folder gets pushed.
```

File: backend/app/utils/fileService.ts [hunk 18]
```diff
@@ -18,17 +18,19 @@ class FileService {
...
 
-    public async convertImage (path: any): Promise<void> {
-        sharp(path).resize(800)
+    public async convertImage (path: any, dimensions: {}): Promise<void> {
+        sharp(path)
+            .resize(dimensions)
     }
 
-    public async storeImage (file: any): Promise<string> {
+    // Types: banner & avatar
```
Comment (AjobK) [inline backend/app/utils/fileService.ts]: More of a question;

Is this comment necessary?

**Review by AjobK — CHANGES_REQUESTED** [review]
Minor things and then merge!

File: .gitignore (hunk 15 above)
Comment (AjobK) [reply .gitignore]:
```suggestion
!backend/app/public/profile/avatar/.gitkeep

```

**Review by AjobK — CHANGES_REQUESTED** [review]

File: backend/app/utils/fileService.ts (hunk 18 above)
Comment (AjobK) [reply backend/app/utils/fileService.ts]:
```suggestion
```

**Review by AjobK — APPROVED** [review]

**Review by ryankroon00 — APPROVED** [review]

**Pull request merged**